File: tslibR/R/decomposition.R
#' STL decomposition
#'
#' Decomposes a time series into trend, seasonal, and remainder components
#' using Seasonal-Trend decomposition with Loess (STL).
#'
#' @param client A [`tslib_client`] object.
#' @param data Numeric vector of observations.
#' @param period Seasonal period (e.g. 12 for monthly data).
#' @param trend_window Optional trend smoother window (must be odd).
#' @param seasonal_window Optional seasonal smoother window (must be odd).
#' @param iterations Optional number of inner robustness iterations.
#' @param outer_iterations Optional number of outer robustness iterations
#'   (set > 0 for robust STL).
#' @return Named list with `trend`, `seasonal`, `remainder`, and
#'   `reconstructed` numeric vectors.
#' @export
tslib_stl <- function(client, data, period,
                      trend_window    = NULL,
                      seasonal_window = NULL,
                      iterations      = NULL,
                      outer_iterations = NULL) {
  body <- .compact(list(
    data            = as.list(data),
    period          = period,
    trendWindow     = trend_window,
    seasonalWindow  = seasonal_window,
    iterations      = iterations,
    outerIterations = outer_iterations
  ))
  .tslib_post(client, "/api/decompose/stl", body)
}

File: tslib-py/tslib/decomposition.py
from typing import List, Optional

from ._http import _Session
from .models import StlResult


class DecompositionAPI:
    def __init__(self, session: _Session) -> None:
        self._s = session

    def stl(
        self,
        data: List[float],
        period: int,
        *,
        trend_window: Optional[int] = None,
        seasonal_window: Optional[int] = None,
        iterations: Optional[int] = None,
        outer_iterations: Optional[int] = None,
    ) -> StlResult:
        body: dict = {"data": data, "period": period}
        if trend_window is not None:
            body["trendWindow"] = trend_window
        if seasonal_window is not None:
            body["seasonalWindow"] = seasonal_window
        if iterations is not None:
            body["iterations"] = iterations
        if outer_iterations is not None:
            body["outerIterations"] = outer_iterations
        return StlResult._from(self._s.post("/api/decompose/stl", body))
